### PR TITLE
Enable easy admin development

### PR DESCRIPTION
**Pull request by soupette**
#### Description of what you did:
This PR allows a user to start a `webpack-dev-server` directly from their application so they can easily customise/develop plugins. Any files added in either `my-app/admin/` or `my-app/extensions/<plugin>/admin` will now be copied and used by the administration panel instead of the original ones.

#### My PR is a:

- [ ] 💥 Breaking change
- [ ] 🐛 Bug fix
- [ ] 💅 Enhancement
- [X] 🚀 New feature

#### Main update on the:

- [X] Admin
- [ ] Documentation
- [ ] Framework
- [ ] Plugin

#### Manual testing done on the following databases:

- [x] Not applicable
- [ ] MongoDB
- [ ] MySQL
- [ ] Postgres
- [ ] SQLite

### CODE REVIEW

File: docs/3.0.0-beta.x/advanced/customize-admin.md [hunk 1]
```diff
@@ -29,9 +29,101 @@ By default, the administration panel is exposed via [http://localhost:1337/admin
...
+
+#### Customizing the `strapi-admin` package
+
+All files added in `my-app/admin/src/` will either be replaced or added
+
+**Example: Changing the available locales of your application**
+
+```bash
+# Create both the admin and admin/src folders
+cd my-app && mkdir admin && mkdir admin/src
```
Comment (alexandrebodin) [inline docs/3.0.0-beta.x/advanced/customize-admin.md]: `mkdir -p admin/src`

File: docs/3.0.0-beta.x/advanced/customize-admin.md [hunk 2]
```diff
@@ -29,9 +29,101 @@ By default, the administration panel is exposed via [http://localhost:1337/admin
...
+# Change the available locales of the administration panel
+touch mkdir/src/i18n.js
```
Comment (alexandrebodin) [inline docs/3.0.0-beta.x/advanced/customize-admin.md]: `touch admin/src/i18n.js`

File: packages/strapi-admin/index.js [hunk 3]
```diff
@@ -152,7 +176,48 @@ async function build({ dir, env, options }) {
...
+
+    console.log(chalk.green('Starting the development server...'));
+    console.log();
+    console.log(
+      chalk.green(`Admin development at http://localhost:${port}/admin`)
+    );
+  });
+}
+
 module.exports = {
```
Comment (alexandrebodin) [inline packages/strapi-admin/index.js]: only build and watch should be exported

File: packages/strapi/lib/commands/build.js [hunk 4]
```diff
@@ -34,6 +34,9 @@ module.exports = async () => {
 
   console.log(`Building your admin UI with ${green(env)} configuration ...`);
 
+  // Create the .cache folder containing the front-end files
+  await strapiAdmin.createCacheDir(dir);
```
Comment (alexandrebodin) [inline packages/strapi/lib/commands/build.js]: should be in the build code

File: packages/strapi/lib/commands/develop.js [hunk 5]
```diff
@@ -7,16 +7,23 @@ const chokidar = require('chokidar');
...
  * `$ strapi develop`
  *
  */
-module.exports = async function({ build }) {
+module.exports = async function({ build, watch }) {
   const dir = process.cwd();
 
-  if (build && !fs.existsSync(path.join(dir, 'build'))) {
+  // Always create the .cache folder containing the front-end files
+  await strapiAdmin.createCacheDir(dir);
```
Comment (alexandrebodin) [inline packages/strapi/lib/commands/develop.js]: should be strapi-admin.build or strapi-admin.watch code not in a command

File: docs/3.0.0-beta.x/cli/CLI.md [hunk 6]
```diff
@@ -37,6 +37,18 @@ Start a Strapi application with autoReload enabled.
...
+```
+strapi develop
+options: [--no-build |--watch ]
+```
+
+- **strapi develop**<br/>
+  Starts your application with the autoReload enabled
+- **strapi develop --no-build**<br/>
+  Starts your application with the autoReload enabled and skip the administration panel build process
+- **strapi develop --watch**<br/>
```
Comment (alexandrebodin) [inline docs/3.0.0-beta.x/cli/CLI.md]: maybe we should rename watch to `--watch-admin`

Comment (alexandrebodin) [reply docs/3.0.0-beta.x/cli/CLI.md]: same for the watch command watch-admin would maybe be clearer

File: packages/strapi-admin/index.js [hunk 7]
```diff
@@ -152,7 +176,48 @@ async function build({ dir, env, options }) {
...
+    env,
+    port,
+    options,
+  };
+
+  const opts = {
+    clientLogLevel: 'silent',
+    hot: true,
+    quiet: true,
+    publicPath: '/admin/',
```
Comment (alexandrebodin) [inline packages/strapi-admin/index.js]: use the options.publicPath you set in the watch command

File: packages/strapi-admin/index.js [hunk 8]
```diff
@@ -152,7 +176,48 @@ async function build({ dir, env, options }) {
...
+    historyApiFallback: {
+      index: '/admin/',
```
Comment (alexandrebodin) [inline packages/strapi-admin/index.js]: same

File: packages/strapi-admin/index.js [hunk 9]
```diff
@@ -152,7 +176,48 @@ async function build({ dir, env, options }) {
   });
 }
 
+async function watch({ dir, port, options }) {
+  const entry = path.join(dir, '.cache', 'admin', 'src', 'app.js');
+  const dest = path.join(dir, 'build');
+  const env = 'development';
+
```
Comment (alexandrebodin) [inline packages/strapi-admin/index.js]: this function should create the cache dir first then start the dev server then start the watcher

File: packages/strapi/lib/commands/watch.js [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    path.join(
+      dir,
+      'extensions',
+      plugin.replace(/^strapi-plugin-/i, ''),
+      'admin'
+    )
+  );
+  const filesToWatch = [admin, ...pluginsToWatch];
+
+  function watchFiles() {
```
Comment (alexandrebodin) [inline packages/strapi/lib/commands/watch.js]: the watcher should be moved into strapi-admin

File: docs/3.0.0-beta.x/advanced/customize-admin.md (hunk 1 above)
Comment (derrickmehaffy) [reply docs/3.0.0-beta.x/advanced/customize-admin.md]: (For clarification mk -p means to make any required parent directories so you can just provide the deepest child directory and any non-existing parents will be created as needed @soupette )

**Pull request merged**